client.py

```diff
@@ -0,0 +1,86 @@
+import socket
+import argparse
+import threading
+import time
+from policy_compiler.helper import safe_open
+
+exit_query="query(exit)\n"
+
+def send_queries(server_address, server_port, workload_file, latency_results):
+  # Open a connection to the server
+  client_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+  client_socket.connect((server_address, server_port))
+
+  # Read the contents of the workload file line by line
+  total_latency = 0
+  request_count = 0
+  with safe_open(workload_file, 'r') as file:
+    for line in file:
+      if line.startswith("#"):
+        continue  # Skip queries starting with '#'
+      # Send each line to the server
+      start_time = time.perf_counter()  # Start the timer
+      client_socket.send(line.encode())
+
+      # Receive the server's response
+      response = client_socket.recv(1024).decode().strip()
+      # print(response)
+      end_time = time.perf_counter()  # End the timer
+
+      # Calculate and accumulate the latency
+      latency = end_time - start_time
+      total_latency += latency
+      request_count += 1
+      # print(request_count)
+  
+  client_socket.send(exit_query.encode())
+  # Close the connection
+  client_socket.close()
+
+  # Save the average latency
+  if request_count > 0:
+    average_latency = total_latency / request_count
+    latency_results.append(average_latency)
+
+def create_client_thread(server_address, server_port, workload_file, latency_results):
+  thread = threading.Thread(target=send_queries, args=(server_address, server_port, workload_file, latency_results))
+  thread.start()
+  return thread
+
+def main():
+  parser = argparse.ArgumentParser(description='Start a client.')
+  parser.add_argument('--workload', help='Path to the workload trace file', default=None, required=True, type=str)
+  parser.add_argument('--address', help='IP address of the server to connect', default="127.0.0.1", required=False, type=str)
+  parser.add_argument('--port', help='Port of the running server to connect', default=1312, required=False, type=int)
+  parser.add_argument('--clients', help='Number of clients to spawn', default=1, type=int)
+  args = parser.parse_args()
+
+  # Start the time measurement before sending the workload
+  start_time = time.perf_counter()
+
+  latency_results = []
+  threads = []
```

Comment:
We can also consider using `multiprocessing` or` concurrent.futures.ProcessPoolExecutor` library as Python threading library can not utilize multiple cores due to GIL ([link](https://towardsdatascience.com/demystifying-python-multiprocessing-and-multithreading-9b62f9875a27)).

Reply:
Legit and useful comment.
I ran some experiments w/o the connection, only with the scanning of the workloads and indeed multiprocessing scaled well while threading did not.
I adapted the client.